ユーザー管理機能

KDItsu opened the pull request:
# What
ユーザー管理機能画面の作成
# Why
ユーザー管理機能を実装するため

 ■必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/57cd7987147091191df9b30055c120d2

■入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/6bb69e625137c660345061072edb31d3

 ■必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/cb2b98a50c61e62094cc9e018f919f32

■ 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/3f1567bc795fcfaa9b36fab0beb5752c

 ■ログアウトに成功する動画
https://gyazo.com/5ff801376fdc36d94af4bf698b1d2af3

 ■テストに成功する画像
https://gyazo.com/79a7060670de5f2b7941a20393d6f4e6

KDItsu committed:
deviseを導入する

KDItsu committed:
ユーザー管理機能テストコード

div-reviewer commented on app/models/user.rb:
```diff
@@ -0,0 +1,14 @@
+  VALID_PASSWORD_REGEX = /\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i.freeze
+  validates :password, format: { with: VALID_PASSWORD_REGEX }
+  validates :nickname, presence: true
+  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
```
全角のバリデーションチャレンジありがとうございました。
現状、先頭文字のみをチェックする記述となります。 氏名のカナのように、`+\z`を使用し、全ての文字をチェックするようにしましょう。

■参考
https://master.tech-camp.in/v2/curriculums/4886#6:~:text=%E3%81%8F%E3%81%A0%E3%81%95%E3%81%84%27%20%7D-,%E5%85%A8%E8%A7%92%E3%81%8B%E3%81%AA/%E3%82%AB%E3%83%8A%E6%BC%A2%E5%AD%97%E3%81%AE%E5%8C%BA%E5%88%A5%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6%E8%80%83%E3%81%88%E3%81%A6%E3%81%BF%E3%82%88%E3%81%86,-%E5%85%A8%E8%A7%92%E3%81%8B%E3%81%AA/%E3%82%AB%E3%83%8A

KDItsu replied:
修正しました。

div-reviewer commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,107 @@
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%# render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
ユーザー登録失敗時にエラーメッセージが正しく表示されない場合、_error_messages.html.erbの呼び出しの記述を見直しましょう。

https://gyazo.com/6bb69e625137c660345061072edb31d3
こちらの撮り直しもお願いいたします。

KDItsu replied:
修正しました。

KDItsu replied:
再提出
■入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/3c92789b7a31b858c7add58e422989ca

div-reviewer commented on app/controllers/application_controller.rb:
```diff
@@ -1,11 +1,11 @@
 class ApplicationController < ActionController::Base
-  before_action :basic_auth
+  before_action :authenticate_user!
```
applicationコントローラーからauthenticate_user!メソッドを削除しましょう。
ユーザー管理機能の実装要件において、未ログイン時のページ遷移制限は求められていないためです。

KDItsu replied:
修正しました。

div-reviewer commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,107 @@
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%# render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

KDItsu replied:
修正しました。

div-reviewer commented on db/migrate/20240503084330_devise_create_users.rb:
```diff
@@ -0,0 +1,51 @@
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,              null: false, unique: true, default: ""
```
deviseがデフォルトでUNIQUE制約を付与しているため、emailカラムに対するunique:trueは削除して問題ございません。

KDItsu replied:
修正しました。

div-reviewer commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              { Faker::Name.initials(number: 2) }
+    email                 { Faker::Internet.email }
+    password              { Faker::Internet.password(min_length: 6) }
```
Fakerのみですと英字のみや数字のみのpasswordが生成されてしまう可能性があります。
password {Fakerによって生成された文字列+'1a'}のように'1a'を足すことで必ず半角英数字混合になるようにしましょう。

KDItsu replied:
修正しました。

div-reviewer commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,119 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password confirmation doesn't match Password")
+      end
+      it 'お名前(全角)は、名字が空では登録できない' do
+        @user.last_name = ''
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Last name can't be blank", "Last name is invalid")
+      end
+      it 'お名前(全角)は、名前が空では登録できない' do
+        @user.first_name = ''
+        @user.valid?
+        expect(@user.errors.full_messages).to include("First name can't be blank", "First name is invalid")
+      end
```
**■テスト全体**
期待するエラーメッセージはitに対して基本1つです。
よって、`("First name can't be blank", "First name is invalid")`のように複数選択しているものに関しては一つだけにしましょう。

KDItsu replied:
修正しました。

div-reviewer reviewed (commented):
KDItsu様、学習お疲れ様です。
ユーザー管理機能の初回レビュー依頼ありがとうございます。プルリクエストの内容を確認させていただきました。

バリデーションやテストの洗い出し、素晴らしくよく出来ていると思います。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。**

KDItsu committed:
バリデーション全ての文字をチェックする

KDItsu committed:
未ログイン時のページ遷移制限は求められていない

KDItsu committed:
デフォルト

deviseがデフォルトでUNIQUE制約を付与しているため、emailカラムに対するunique:trueは削除

KDItsu committed:
Faker

Fakerのみですと英字のみや数字のみのpasswordが生成されてしまう可能性があります。
password {Fakerによって生成された文字列+'1a'}のように'1a'を足すことで必ず半角英数字混合になるようにしましょう。

KDItsu committed:
エラーメッセージ

期待するエラーメッセージはitに対して基本1つです。
よって、("First name can't be blank", "First name is invalid")のように複数選択しているものに関しては一つだけにしましょう。

KDItsu committed:
エラーメッセージ

ユーザー登録失敗時にエラーメッセージが正しく表示されない場合、_error_messages.html.erbの呼び出しの記述を見直しましょう。

maimai0401 commented on app/models/user.rb:
```diff
@@ -0,0 +1,14 @@
+  validates :password, format: { with: VALID_PASSWORD_REGEX }
+  validates :nickname, presence: true
+  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
+  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
```
```
  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
```



これらの記述ですが、/\A[ぁ-んァ-ヶ一-龥々ー]+\z/と修正しておきましょう。

**理由**
・現在の記述では、「佐々木」のような「々」や「マーヴェリック」のような「ヴ」「ー」が許可されていない正規表現となっているため
・2文字目以降に全角漢字・ひらがな・カタカナ以外が入力された時にバリデーションが思った通りに動かない可能性があるため

【補足】 ・長音記号(ー)も考慮（龥々ー） ・文字列の最後まで考慮（+\z）

**■参考**
```ruby
コース：最終課題
タイトル：最終課題補足カリキュラム〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜

バリデーションを設定しましょう
```

KDItsu replied:
修正しました。

maimai0401 commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,107 @@
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

maimai0401 commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,107 @@
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%# render 'shared/error_messages', model: f.object %>
```
登録に失敗した際にエラーメッセージが表示されるように修正しましょう。

KDItsu replied:
修正しました。
ありがとうございます。
再々提出
■入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/b0fe78666a879e1ea2e566aed87fc155

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください。
どうぞよろしくおねがいします。

KDItsu committed:
バリデーション１

validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/ }
これらの記述ですが、/\A[ぁ-んァ-ヶ一-龥々ー]+\z/と修正しておきましょう。

理由
・現在の記述では、「佐々木」のような「々」や「マーヴェリック」のような「ヴ」「ー」が許可されていない正規表現となっているため
・2文字目以降に全角漢字・ひらがな・カタカナ以外が入力された時にバリデーションが思った通りに動かない可能性があるため

【補足】 ・長音記号(ー)も考慮（龥々ー） ・文字列の最後まで考慮（+\z）

KDItsu committed:
エラーメッセージ

maimai0401 commented:
レビュー依頼ありがとうございます。

こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【注意】
```ruby
①（次のブランチを切る前に）プルリクエストのMergeとmainブランチへのPullを忘れないようにしましょう。
②次のフェーズに進む際は必ず新しいブランチを作ってから作業しましょう。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
引き続き頑張りましょう！応援しております！
